### PR TITLE
Routed subinterface enhancements

### PR DESCRIPTION
**Pull request by preetham-singh**
What I did
Routed subinterfae enhancements HLD [#833](https://github.com/Azure/SONiC/pull/833)
Add support for long name and short name routed subinterfaces.
Add support for long name and short name routed subinterfaces.
This swss common library provides APIS for:
- Subinterface validation checks
- Get parent interface corresponding to subinterface(short/long name)
- Get subinterface index
 
Why I did it
Routed subinterface feature was broken for physical and portchannel subinterfaces for subinterface name exceeding 15 characters due to kernel limitation of netdev name length of 15.

How I verified it
Routed subinterface Unit tests

### CODE REVIEW

File: common/ifcommon.cpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+bool subIntf::isShortName()
+{
+    if (isCompressed == true)
+    {
+        return true;
+    }
+    else
+    {
+        return false;
+    }
```
Comment (kcudnik) [inline common/ifcommon.cpp:119]: this could be single line return iscommpressed == true;

File: common/ifcommon.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        public:
+            subIntf(const std::string &ifName);
+            bool isValid();
+            std::string parentIntf();
+            int subIntfIdx();
+            std::string longName();
+            std::string shortName();
+            bool isShortName();
+        private:
+#define VLAN_SUB_INTERFACE_SEPARATOR   "."
```
Comment (kcudnik) [inline common/ifcommon.h:16]: define is not bound by class or namespace, ether put it on top of the file (and also use pragma once) or make it as constexpr

File: common/ifcommon.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+            std::string alias = "";
+            std::string subIfIdx = "";
+            std::string parentIf = "";
+            std::string parentIfShortName = "";
```
Comment (kcudnik) [inline common/ifcommon.h:20]: initialization strings like this is not required, they by default are empty

File: common/ifcommon.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    class subIntf
+    {
+        public:
+            subIntf(const std::string &ifName);
+            bool isValid();
+            std::string parentIntf();
+            int subIntfIdx();
+            std::string longName();
+            std::string shortName();
+            bool isShortName();
```
Comment (kcudnik) [inline common/ifcommon.h:14]: all those functions should be const

File: common/ifcommon.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
+#ifndef __IFCOMMON_H__
+#define __IFCOMMON_H__
+
+namespace swss {
+    class subIntf
```
Comment (kcudnik) [inline common/ifcommon.h:5]: preffered cilename should be as class it defines so like subintf.h

File: common/ifcommon.cpp (hunk 1 above)
Comment (preetham-singh) [reply common/ifcommon.cpp]: Updated as part of next commit.

File: common/ifcommon.h (hunk 3 above)
Comment (preetham-singh) [reply common/ifcommon.h]: Updated as part of next commit.

File: common/ifcommon.h (hunk 5 above)
Comment (preetham-singh) [reply common/ifcommon.h]: Updated file names.

**Review by kcudnik — APPROVED** [review]

**Comment by lguohan** [comment]
@prsunny , why is this pr on swss-common? are you sponsoring this pr?

**Comment by preetham-singh** [comment]
Closing this pull request since this code is moved to swss/lib which can be found as part of swss PR [2017](https://github.com/Azure/sonic-swss/pull/2017)

**Pull request closed without merging**